usr.bin/wc/wc.1
.\"
.\" SPDX-License-Identifier: BSD-3-Clause
.\"
.\" Copyright (c) 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" This code is derived from software contributed to Berkeley by
.\" the Institute of Electrical and Electronics Engineers, Inc.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.Dd July 16, 2025
.Dt WC 1
.Os
.Sh NAME
.Nm wc
.Nd word, line, character, and byte count
.Sh SYNOPSIS
.Nm
.Op Fl -libxo
.Op Fl c | Fl m
.Op Fl Llw
.Op Ar
.Sh DESCRIPTION
The
.Nm
utility displays the number of lines, words, and bytes contained in each
input
.Ar file ,
or standard input (if no file is specified) to the standard output.
A line is defined as a string of characters delimited by a
.Dq newline
character.
Characters beyond the final
.Dq newline
character will not be included
in the line count.
.Pp
A word is defined as a string of characters delimited by white space
characters.
White space characters are the set of characters for which the
.Xr iswspace 3
function returns true.
If more than one input file is specified, a line of cumulative counts
for all the files is displayed on a separate line after the output for
the last file.
.Pp
The following options are available:
.Bl -tag -width "--libxo"
.It Fl -libxo
Generate output via
.Xr libxo 3
in a selection of different human and machine readable formats.
See
.Xr xo_options 7
for details on command line arguments.
.It Fl L
Write the length of the line containing the most bytes (default) or characters
(when
.Fl m
is provided)
to standard output.
When more than one
.Ar file
argument is specified, the longest input line of
.Em all
files is reported as the value of the final
.Dq total .
.It Fl c
The number of bytes in each input file
is written to the standard output.
This will cancel out any prior usage of the
.Fl m
option.
.It Fl l
The number of lines in each input file
is written to the standard output.
.It Fl m
The number of characters in each input file is written to the standard output.
If the current locale does not support multibyte characters, this
is equivalent to the
.Fl c
option.
This will cancel out any prior usage of the
.Fl c
option.
.It Fl w
The number of words in each input file
is written to the standard output.
.El
.Pp
When an option is specified,
.Nm
only reports the information requested by that option.
The order of output always takes the form of line, word,
byte, and file name.
The default action is equivalent to specifying the
.Fl c , l
and
.Fl w
options.
.Pp
If no files are specified, the standard input is used and no
file name is displayed.
The prompt will accept input until receiving EOF, or
.Bq ^D
in most environments.
.Pp
If
.Nm
receives a
.Dv SIGINFO
(see the
.Cm status
argument for
.Xr stty 1 )
signal, the interim data will be written
to the standard error output in the same format
as the standard completion message.
.Sh ENVIRONMENT
The
.Ev LANG , LC_ALL
and
.Ev LC_CTYPE
environment variables affect the execution of
.Nm
as described in
.Xr environ 7 .
.Sh EXIT STATUS
.Ex -std
.Sh EXAMPLES
Count the number of characters, words and lines in each of the files
.Pa report1
and
.Pa report2
as well as the totals for both:
.Pp
.Dl "wc -mlw report1 report2"
.Pp
Find the longest line in a list of files:
.Pp
.Dl "wc -L file1 file2 file3 | fgrep total"
.Sh COMPATIBILITY
Historically, the
.Nm
utility was documented to define a word as a
.Do
maximal string of
characters delimited by <space>, <tab> or <newline> characters
.Dc .
The implementation, however, did not handle non-printing characters
correctly so that
.Dq Li "\ \ ^D^E\ \ "
counted as 6 spaces, while
.Dq Li foo^D^Ebar
counted as 8 characters.
.Bx 4
systems after
.Bx 4.3
modified the implementation to be consistent
with the documentation.
This implementation defines a
.Dq word
in terms of the
.Xr iswspace 3
function, as required by
.St -p1003.2 .
.Pp
The
.Fl L
option is a non-standard
.Fx
extension, compatible with the
.Fl L
option of the GNU
.Nm
utility.
.Sh SEE ALSO
.Xr iswspace 3 ,
.Xr libxo 3 ,
.Xr xo_options 7
.Sh STANDARDS
The
.Nm
utility conforms to
.St -p1003.1-2001 .
.Sh HISTORY
A
.Nm
command appeared in
.At v1 .


NEWS
-------------------------------------------------------------------------------
-- Copyright 2018-2024,2025 Thomas E. Dickey                                 --
-- Copyright 1998-2017,2018 Free Software Foundation, Inc.                   --
--                                                                           --
-- Permission is hereby granted, free of charge, to any person obtaining a   --
-- copy of this software and associated documentation files (the             --
-- "Software"), to deal in the Software without restriction, including       --
-- without limitation the rights to use, copy, modify, merge, publish,       --
-- distribute, distribute with modifications, sublicense, and/or sell copies --
-- of the Software, and to permit persons to whom the Software is furnished  --
-- to do so, subject to the following conditions:                            --
--                                                                           --
-- The above copyright notice and this permission notice shall be included   --
-- in all copies or substantial portions of the Software.                    --
--                                                                           --
-- THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS   --
-- OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF                --
-- MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN --
-- NO EVENT SHALL THE ABOVE COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,       --
-- DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR     --
-- OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE --
-- USE OR OTHER DEALINGS IN THE SOFTWARE.                                    --
--                                                                           --
-- Except as contained in this notice, the name(s) of the above copyright    --
-- holders shall not be used in advertising or otherwise to promote the      --
-- sale, use or other dealings in this Software without prior written        --
-- authorization.                                                            --
-------------------------------------------------------------------------------
-- $Id: NEWS,v 1.4390 2025/12/30 20:31:07 tom Exp $
-------------------------------------------------------------------------------

This is a log of changes that ncurses has gone through since Zeyd started
working with Pavel Curtis' original work, pcurses, in 1992.

Changes through 1.9.9e are recorded by Zeyd M Ben-Halim.
Changes since 1.9.9e are recorded by Thomas E Dickey.

Contributors include those who have provided patches (even small ones), as well
as those who provide useful information (bug reports, analyses).  Changes with
no cited author are the work of Thomas E Dickey (TD).

A few contributors may be cited in this file by their initials.
Each accounts for half of one percent or more of the changes since 1.9.9e.
See the AUTHORS file for the corresponding full names.

Changes through 1.9.9e did not credit all contributions;
it is not possible to add this information.

20251230 6.6 release for upload to ftp.gnu.org
	+ update announcement
	+ corrected an ifdef needed for mouse support in MinGW/Windows
	+ eliminate remaining duplicate code between MinGW/Windows drivers

20251227
	+ make win32_curses.h obsolete in favor of nc_win32.h
	+ modify MinGW32 configuration to account for its use of Windows-style
	  pathnames in filesystem checks.
	+ replace --enable-exp-win32 option with --enable-named-pipes

20251220
	> in-progress work to merge MinGW/Windows port.
	  + eliminate EXP_WIN32_DRIVER with USE_NAMED_PIPES
	  + change MS_TERMINAL to DEFAULT_TERM_VAR

20251213
	+ in-progress work to merge MinGW/Windows port.
	+ add a null-pointer check in tic -c option.
	+ add a limit-check in infocmp -i option (report/example by Yixuan Cao).

20251206
	+ in-progress work to merge MinGW/Windows port.

20251129
	+ in-progress work to merge MinGW/Windows port.

20251123
	+ revert change using NCURSES_SBOOL in tic.h which interfered with
	  sign-extension in tigetflag (Debian #1121191, cf:  20251101).

20251122
	+ in-progress work to merge MinGW/Windows port.
	+ correct expression for TINFO_LIB variable in misc/ncurses-config.in
	  (Debian #1121208, cf: 20251004).

20251115
	+ check for special case of wcrtomb() converting a single byte code to
	  a different single byte code, which glibc does for code 160 in the
	  KOI8-R encoding (report by Yury V Zaytsev).
	+ improve formatting/style of manpages (patches by Branden Robinson).
	+ amend recent changes to mouse handling to avoid a case which caused
	  a loop in wgetch until additional input is provided (htop #1769,
	  cf: 20250913).

20251101
	+ adjust generated terminfo.5 to omit tbl "x" column modifier if that
	  is unsupported, i.e., with old nroff/tbl (adapted from patch by
	  Branden Robinson).
	+ in-progress work to merge MinGW/Windows port.

20251025
	+ amend locale-portability fix in tty_update.c (MidnightCommander #4799,
	  cf: 20240203).
	+ add testing utility report_ctype
	+ update configure macros, improving check for $host_os
	+ correct a recent change to curs_termino.3x, along with formatting
	  fixes for manpages.
	+ corrected note about box() in curs_border.3x (patch by Jakub Horky,
	  cf: 20250201).

20251018
	+ add check for build-time utilities, in case cross-compiling is setup
	  with an invalid $BUILD_CC (prompted by discussion with Stas Sergeev).
	+ modify configure-script and makefiles to make it easier to override
	  the "--with-install-prefix" default with environment variable
	  (report by Stas Sergeev).

20251010
	+ fix stray "\" in curs_extend.3x, regen html manpages.
	+ regen html manpages with man2html 1.42, fixing a few cases where
	  headers were not linked (report by Branden Robinson).

20251004
	+ improve formatting/style of manpages (patches by Branden Robinson).
	+ in-progress work to merge MinGW/Windows port.

usr.bin/wc/wc.c
/*-
 * SPDX-License-Identifier: BSD-3-Clause
 *
 * Copyright (c) 1980, 1987, 1991, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/capsicum.h>
#include <sys/param.h>
#include <sys/stat.h>

#include <capsicum_helpers.h>
#include <ctype.h>
#include <errno.h>
#include <fcntl.h>
#include <locale.h>
#include <stdbool.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <wchar.h>
#include <wctype.h>
#include <libxo/xo.h>

#include <libcasper.h>
#include <casper/cap_fileargs.h>

#define WC_XO_VERSION "1"

static const char *stdin_filename = "stdin";

static fileargs_t *fa;
static uintmax_t tlinect, twordct, tcharct, tlongline;
static bool doline, doword, dochar, domulti, dolongline;
static volatile sig_atomic_t siginfo;
static xo_handle_t *stderr_handle;

static void	show_cnt(const char *file, uintmax_t linect, uintmax_t wordct,
		    uintmax_t charct, uintmax_t llct);
static int	cnt(const char *);
static void	usage(void);

static void
siginfo_handler(int sig __unused)
{

	siginfo = 1;
}

static void
reset_siginfo(void)
{

	signal(SIGINFO, SIG_DFL);
	siginfo = 0;
}

int
main(int argc, char *argv[])
{
	int ch, errors, total;
	cap_rights_t rights;

	(void) setlocale(LC_CTYPE, "");

	argc = xo_parse_args(argc, argv);
	if (argc < 0)
		exit(EXIT_FAILURE);

	while ((ch = getopt(argc, argv, "clmwL")) != -1)
		switch((char)ch) {
		case 'l':
			doline = true;
			break;
		case 'w':
			doword = true;
			break;
		case 'c':
			dochar = true;
			domulti = false;
			break;
		case 'L':
			dolongline = true;
			break;
		case 'm':
			domulti = true;
			dochar = false;
			break;
		case '?':
		default:
			usage();
		}
	argv += optind;
	argc -= optind;

	fa = fileargs_init(argc, argv, O_RDONLY, 0,
	    cap_rights_init(&rights, CAP_READ, CAP_FSTAT), FA_OPEN);
	if (fa == NULL)
		xo_err(EXIT_FAILURE, "Unable to initialize casper");
	caph_cache_catpages();
	if (caph_limit_stdio() < 0)
		xo_err(EXIT_FAILURE, "Unable to limit stdio");
	if (caph_enter_casper() < 0)
		xo_err(EXIT_FAILURE, "Unable to enter capability mode");

	/* Wc's flags are on by default. */
	if (!(doline || doword || dochar || domulti || dolongline))
		doline = doword = dochar = true;

	stderr_handle = xo_create_to_file(stderr, XO_STYLE_TEXT, 0);

	xo_set_version(WC_XO_VERSION);
	xo_open_container("wc");
	xo_open_list("file");

	(void)signal(SIGINFO, siginfo_handler);
	errors = 0;
	total = 0;
	if (argc == 0) {
		xo_open_instance("file");
		if (cnt(NULL) != 0)
			++errors;
		xo_close_instance("file");
	} else {
		while (argc--) {
			xo_open_instance("file");
			if (cnt(*argv++) != 0)
				++errors;
			xo_close_instance("file");
			++total;
		}
	}

	xo_close_list("file");

	if (total > 1) {
		xo_open_container("total");
		show_cnt("total", tlinect, twordct, tcharct, tlongline);
		xo_close_container("total");
	}

	fileargs_free(fa);
	xo_close_container("wc");
	if (xo_finish() < 0)
		xo_err(EXIT_FAILURE, "stdout");
	exit(errors == 0 ? EXIT_SUCCESS : EXIT_FAILURE);
}

static void
show_cnt(const char *file, uintmax_t linect, uintmax_t wordct,
    uintmax_t charct, uintmax_t llct)
{
	xo_handle_t *xop;

	if (!siginfo)
		xop = NULL;
	else {
		xop = stderr_handle;
		siginfo = 0;
	}

	if (doline)
		xo_emit_h(xop, " {:lines/%7ju/%ju}", linect);
	if (doword)
		xo_emit_h(xop, " {:words/%7ju/%ju}", wordct);
	if (dochar || domulti)
		xo_emit_h(xop, " {:characters/%7ju/%ju}", charct);
	if (dolongline)
		xo_emit_h(xop, " {:long-lines/%7ju/%ju}", llct);
	if (file != stdin_filename)
		xo_emit_h(xop, " {:filename/%s}\n", file);
	else
		xo_emit_h(xop, "\n");
}

static int
cnt(const char *file)
{
	static char buf[MAXBSIZE];
	struct stat sb;
	mbstate_t mbs;
	const char *p;
	uintmax_t linect, wordct, charct, llct, tmpll;
	ssize_t len;
	size_t clen;
	int fd;
	wchar_t wch;
	bool gotsp, warned;

	linect = wordct = charct = llct = tmpll = 0;
	if (file == NULL) {
		fd = STDIN_FILENO;
		file = stdin_filename;
	} else if ((fd = fileargs_open(fa, file)) < 0) {
		xo_warn("%s: open", file);
		return (1);
	}
	if (doword || (domulti && MB_CUR_MAX != 1))
		goto word;
	/*
	 * If all we need is the number of characters and it's a regular file,
	 * just stat it.
	 */
	if (doline == 0 && dolongline == 0) {
		if (fstat(fd, &sb)) {
			xo_warn("%s: fstat", file);
			(void)close(fd);
			return (1);
		}
		/* pseudo-filesystems advertize a zero size */
		if (S_ISREG(sb.st_mode) && sb.st_size > 0) {
			reset_siginfo();
			charct = sb.st_size;
			show_cnt(file, linect, wordct, charct, llct);
			tcharct += charct;
			(void)close(fd);
			return (0);
		}
	}
	/*
	 * For files we can't stat, or if we need line counting, slurp the
	 * file.  Line counting is split out because it's a lot faster to get
	 * lines than to get words, since the word count requires locale
	 * handling.
	 */
	while ((len = read(fd, buf, sizeof(buf))) != 0) {
		if (len < 0) {
			xo_warn("%s: read", file);
			(void)close(fd);
			return (1);
		}
		if (siginfo)
			show_cnt(file, linect, wordct, charct, llct);
		charct += len;
		if (doline || dolongline) {
			for (p = buf; len > 0; --len, ++p) {
				if (*p == '\n') {
					if (tmpll > llct)
						llct = tmpll;
					tmpll = 0;
					++linect;
				} else {
					tmpll++;
				}
			}
		}
	}
	reset_siginfo();
	if (doline)
		tlinect += linect;
	if (dochar)
		tcharct += charct;
	if (dolongline && llct > tlongline)
		tlongline = llct;
	show_cnt(file, linect, wordct, charct, llct);
	(void)close(fd);
	return (0);

	/* Do it the hard way... */
word:	gotsp = true;
	warned = false;
	memset(&mbs, 0, sizeof(mbs));
	while ((len = read(fd, buf, sizeof(buf))) != 0) {
		if (len < 0) {
			xo_warn("%s: read", file);
			(void)close(fd);
			return (1);
		}
		p = buf;
		while (len > 0) {
			if (siginfo)
				show_cnt(file, linect, wordct, charct, llct);
			if (!domulti || MB_CUR_MAX == 1) {
				clen = 1;
				wch = (unsigned char)*p;
			} else if ((clen = mbrtowc(&wch, p, len, &mbs)) == 0) {
				clen = 1;
			} else if (clen == (size_t)-1) {
				if (!warned) {
					errno = EILSEQ;
					xo_warn("%s", file);
					warned = true;
				}
				memset(&mbs, 0, sizeof(mbs));
				clen = 1;
				wch = (unsigned char)*p;
			} else if (clen == (size_t)-2) {
				break;
			}
			charct++;
			if (wch != L'\n')
				tmpll++;
			len -= clen;
			p += clen;
			if (wch == L'\n') {
				if (tmpll > llct)
					llct = tmpll;
				tmpll = 0;
				++linect;
			}
			if (iswspace(wch)) {
				gotsp = true;
			} else if (gotsp) {
				gotsp = false;
				++wordct;
			}
		}
	}
	reset_siginfo();
	if (domulti && MB_CUR_MAX > 1) {
		if (mbrtowc(NULL, NULL, 0, &mbs) == (size_t)-1 && !warned)
			xo_warn("%s", file);
	}
	if (doline)
		tlinect += linect;
	if (doword)
		twordct += wordct;
	if (dochar || domulti)
		tcharct += charct;
	if (dolongline && llct > tlongline)
		tlongline = llct;
	show_cnt(file, linect, wordct, charct, llct);
	(void)close(fd);
	return (0);
}

static void
usage(void)
{
	xo_error("usage: wc [-c | -m] [-Llw] [file ...]\n");
	exit(EXIT_FAILURE);
}
